apps/server/src/middleware/authenticate.ts

```diff
@@ -0,0 +1,125 @@
+
+// verify password and set cookies + redirect appropriately
+loginRouter.post('/', (req, res) => {
+  res.clearCookie('token');
+  const { password: reqPassword, redirect } = req.body;
+
+  if (!password || (reqPassword && hashPassword(reqPassword) === hashedPassword)) {
+    res.cookie('token', hashedPassword, {
+      httpOnly: false, // allow websocket to access cookie
+      secure: true,
+      path: '/', // allow cookie to be accessed from any path
+      sameSite: 'strict',
+    });
+    res.redirect(redirect || '/');
+    return;
+  }
+
+  res.status(401).send('Unauthorized');
+});
+
+/**
+ * Express middleware to authenticate requests
+ * @param {string} prefix - Prefix is used for the client hashes in Ontime Cloud
+ */
+export function makeAuthenticateMiddleware(prefix: string) {
+  // we dont need to initialise the authenticate middleware if there is no password
+  if (!hasPassword) {
+    return { authenticate: noopMiddleware, authenticateAndRedirect: noopMiddleware };
+  }
+
+  function authenticate(req: Request, res: Response, next: NextFunction) {
+    if (!password) {
+      return next();
+    }
+
+    const token = req.query.token || req.cookies?.token;
+    if (token && token === hashedPassword) {
+      return next();
+    }
+
+    res.status(401).send('Unauthorized');
+  }
+
+  function authenticateAndRedirect(req: Request, res: Response, next: NextFunction) {
+    if (!password) {
+      next();
+    }
+
+    // we expect the token to be in the cookies
+    if (req.cookies?.token === hashedPassword) {
+      return next();
+    }
+
+    // we use query params for generating authenticated URLs and for clients like the companion module
+    // if the user gives is a token in the query params, we set the cookie to be used in further requests
+    if (req.query.token === hashedPassword) {
+      res.cookie('token', hashedPassword, {
+        httpOnly: false, // allow websocket to access cookie
+        secure: true,
+        path: '/', // allow cookie to be accessed from any path
+        sameSite: 'strict',
+      });
+      return next();
+    }
+
+    // Allow access to specific public assets without authentication
+    if (publicAssets.includes(req.path)) {
+      return next();
+    }
+
+    const redirect = req.originalUrl.startsWith('/login')
+      ? `${prefix}/login`
+      : `${prefix}/login?redirect=${req.originalUrl}`;
+
+    res.redirect(redirect);
+  }
+
+  return { authenticate, authenticateAndRedirect };
+}
+
+/**
+ * Middleware to authenticate a WebSocket connection with a token in the cookie
+ */
+export function authenticateSocket(_ws: WebSocket, req: IncomingMessage, next: (error?: Error) => void) {
+  if (!hasPassword) {
+    return next();
+  }
+
+  const cookieString = req.headers.cookie;
+  if (typeof cookieString === 'string') {
+    const [key, value] = cookieString.split('=');
+    if (key === 'token' && value === hashedPassword) {
+      return next();
+    }
+  }
```

Comment:
```suggestion
  }
  
  const url = new URL(`http://localhost${req.url}`);
  const maybeToken = url.searchParams.get('token');
  if (maybeToken === hashedPassword) {
    return next();
  }
```

allow ws clients without cookies to connect

Reply:
how would the token come up here? do we need an interface in ontime to generate the token or should the websocket client include the raw password in the connection?
What is the normal use case?

Reply:
I was thinking for when companion connects to the ws endpoint directly 

Yeah some place to get the token would be needed, but that could be in the dashboard